File: frontend/types/financial-health.ts
export type HealthCardTone = 'good' | 'warning' | 'danger';

export type SavingsRateZone = 'weak' | 'normal' | 'good';
export type DtiZone = 'normal' | 'acceptable' | 'dangerous' | 'critical';
export type LeverageZone = 'normal' | 'moderate' | 'critical';
export type DisciplineZone = 'weak' | 'medium' | 'good' | 'excellent';
export type FiZone = 'dependent' | 'partial' | 'on_way' | 'free';
export type FiScoreZone = 'start' | 'growth' | 'on_way' | 'freedom';

export type ChronicViolation = {
  category_name: string;
  months_count: number;
  overage_percent: number;
};

export type ChronicUnderperformer = {
  category_id: number;
  category_name: string;
  direction: string;
  direction_label: string;
  months_count: number;
  avg_fulfillment: number;
  trend: string;
  last_planned: number;
  last_actual: number;
};

export type UnplannedCategory = {
  category_id: number;
  category_name: string;
  direction: string;
  direction_label: string;
  avg_monthly_amount: number;
  months_with_spending: number;
};

export type DisciplineHistoryPoint = {
  month: string;
  value: number;
};

export type DirectionHeatmapRow = {
  direction: string;
  label: string;
  planned: number;
  actual: number;
  fulfillment: number;
};

export type MonthlyHealthSnapshot = {
  month: string;
  label: string;
  income: number;
  essential: number;
  secondary: number;
  planned_income: number;
  actual_income: number;
  planned_expenses: number;
  actual_expenses: number;
  savings: number;
  savings_rate: number;
  essential_rate: number;
  secondary_rate: number;
  dti: number;
  fi_score: number;
  discipline: number | null;
  direction_heatmap: DirectionHeatmapRow[];
};

export type FIScoreHistory = {
  current: number;
  previous: number;
  baseline: number;
};

// FI-score v1.4 (Phase 4, 2026-04-19): 4 components, weights 0.20+0.30+0.25+0.25
export type FIScoreComponents = {
  savings_rate: number;      // weight 0.20
  capital_trend: number;     // weight 0.30 — capital trajectory
  dti_inverse: number;       // weight 0.25
  buffer_stability: number;  // weight 0.25 — deposit months / 6 * 10
  months_calculated?: number;
  history?: FIScoreHistory;
};

export type FinancialHealth = {
  savings_rate: number;
  avg_savings_rate: number;
  savings_rate_zone: SavingsRateZone;
  monthly_avg_balance: number;
  months_calculated: number;
  daily_limit: number;
  daily_limit_with_carry: number;
  carry_over_days: number;
  dti: number;
  dti_zone: DtiZone;
  dti_total_payments: number;
  dti_income: number;
  leverage: number;
  leverage_zone: LeverageZone;
  leverage_total_debt: number;
  leverage_own_capital: number;
  real_assets_total?: number;
  discipline: number | null;
  discipline_zone: DisciplineZone | null;
  discipline_violations: ChronicViolation[];
  chronic_underperformers: ChronicUnderperformer[];
  unplanned_categories: UnplannedCategory[];
  discipline_history?: DisciplineHistoryPoint[];
  fi_percent: number;
  fi_zone: FiZone;
  fi_capital_needed: number;
  fi_passive_income: number;
  fi_monthly_gap: number;
  avg_monthly_expenses?: number;
  fi_score: number;
  fi_score_zone: FiScoreZone;
  fi_score_components: FIScoreComponents;
  monthly_history: MonthlyHealthSnapshot[];
};

export type RealAssetType = 'real_estate' | 'car' | 'other';

export type RealAsset = {
  id: number;
  asset_type: RealAssetType;
  name: string;
  estimated_value: number;
  linked_account_id: number | null;
  updated_at: string;
};

export type RealAssetPayload = {
  asset_type: RealAssetType;
  name: string;
  estimated_value: number;
  linked_account_id?: number | null;
};

File: app/schemas/financial_health.py
from __future__ import annotations

from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, ConfigDict, Field


class RealAssetCreate(BaseModel):
    asset_type: str = Field(max_length=32)
    name: str = Field(max_length=255)
    estimated_value: Decimal = Field(ge=0)
    linked_account_id: int | None = None


class RealAssetUpdate(BaseModel):
    asset_type: str | None = Field(default=None, max_length=32)
    name: str | None = Field(default=None, max_length=255)
    estimated_value: Decimal | None = Field(default=None, ge=0)
    linked_account_id: int | None = None


class RealAssetResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    asset_type: str
    name: str
    estimated_value: Decimal
    linked_account_id: int | None
    updated_at: datetime


class CapitalHistoryPoint(BaseModel):
    month: str
    label: str
    liquid: float
    deposit: float
    broker: float
    receivable: float
    real_assets: float
    credit_debt: float
    counterparty_debt: float
    total_debt: float
    liquid_capital: float
    net_capital: float


class ChronicViolation(BaseModel):
    category_name: str
    months_count: int
    overage_percent: float


class ChronicUnderperformer(BaseModel):
    category_id: int
    category_name: str
    direction: str
    direction_label: str
    months_count: int
    avg_fulfillment: float
    trend: str
    last_planned: float
    last_actual: float


class UnplannedCategory(BaseModel):
    category_id: int
    category_name: str
    direction: str
    direction_label: str
    avg_monthly_amount: float
    months_with_spending: int


class FIScoreHistory(BaseModel):
    current: float
    previous: float
    baseline: float


class FIScoreComponents(BaseModel):
    # v1.4 weights (2026-04-19): 0.20 + 0.30 + 0.25 + 0.25 = 1.00
    savings_rate: float        # weight 0.20 — normalised basic flow / regular income
    capital_trend: float       # weight 0.30 — capital trajectory (3-month trend)
    dti_inverse: float         # weight 0.25 — 10 - DTI%/6
    buffer_stability: float    # weight 0.25 — deposit months / 6 * 10
    months_calculated: int | None = None
    history: FIScoreHistory | None = None


class DirectionHeatmapRow(BaseModel):
    direction: str
    label: str
    planned: float
    actual: float
    fulfillment: float


class MonthlyHealthSnapshot(BaseModel):
    month: str
    label: str
    income: float
    essential: float
    secondary: float
    planned_income: float
    actual_income: float
    planned_expenses: float
    actual_expenses: float
    savings: float
    savings_rate: float
    essential_rate: float
    secondary_rate: float
    dti: float
    fi_score: float
    discipline: float | None
    direction_heatmap: list[DirectionHeatmapRow]


class FinancialHealthResponse(BaseModel):
    savings_rate: float
    avg_savings_rate: float = 0.0
    savings_rate_zone: str

    monthly_avg_balance: float
    months_calculated: int

    daily_limit: float
    daily_limit_with_carry: float
    carry_over_days: float

    dti: float
    dti_zone: str
    dti_total_payments: float
    dti_income: float

    leverage: float
    leverage_zone: str
    leverage_total_debt: float
    leverage_own_capital: float
    real_assets_total: float = 0.0

    discipline: float | None
    discipline_zone: str | None
    discipline_violations: list[ChronicViolation]
    chronic_underperformers: list[ChronicUnderperformer] = []
    unplanned_categories: list[UnplannedCategory] = []

    fi_percent: float
    fi_zone: str
    fi_capital_needed: float
    fi_passive_income: float
    fi_monthly_gap: float = 0.0

    fi_score: float
    fi_score_zone: str
    fi_score_components: FIScoreComponents
    monthly_history: list[MonthlyHealthSnapshot] = []
